rust/azure_iot_operations_services/src/leased_lock/lock.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Client for Lock operations.
+
+use std::{sync::Arc, time::Duration};
+
+use crate::leased_lock::{Error, ErrorKind, lease::Client as LeaseClient};
+use crate::state_store::{self};
+use azure_iot_operations_mqtt::interface::ManagedClient;
+use azure_iot_operations_protocol::common::hybrid_logical_clock::HybridLogicalClock;
+
+/// Lock client struct.
+pub struct Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    lease_client: LeaseClient<C>,
+}
+
+/// Lock client implementation
+///
+/// Notes:
+/// Do not call any of the methods of this client after the `state_store` parameter is shutdown.
+/// Calling any of the methods in this implementation after the `state_store` is shutdown results in undefined behavior.
+impl<C> Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync,
+    C::PubReceiver: Send + Sync,
+{
+    /// Create a new Lock Client.
+    ///
+    /// Notes:
+    /// - `lock_holder_name` is expected to be the client ID used in the underlying MQTT connection settings.
+    /// - There must be one instance of `lock::Client` per lock.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`LockNameLengthZero`](ErrorKind::LockNameLengthZero) if the `lock_name` is empty
+    ///
+    /// [`struct@Error`] of kind [`LockHolderNameLengthZero`](ErrorKind::LockHolderNameLengthZero) if the `lock_holder_name` is empty
+    pub fn new(
+        state_store: Arc<state_store::Client<C>>,
+        lock_name: Vec<u8>,
+        lock_holder_name: Vec<u8>,
+    ) -> Result<Self, Error> {
+        if lock_name.is_empty() {
+            return Err(Error(ErrorKind::LockNameLengthZero));
+        }
+
+        if lock_holder_name.is_empty() {
+            return Err(Error(ErrorKind::LockHolderNameLengthZero));
+        }
+
+        let lease_client = LeaseClient::new(state_store, lock_name, lock_holder_name)?;
```

Comment:
any errors that can be returned from this `?` that should be in the doc?

Reply:
No, the only errors are from key or holder being empty, and they are covered already.